rust/server/src/vss_service.rs

```diff
@@ -0,0 +1,154 @@
+	PutObjectResponse,
+};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+
+#[derive(Clone)]
+pub struct VssService {
+	store: Arc<dyn KvStore>,
+	authorizer: Arc<dyn Authorizer>,
+}
+
+impl VssService {
+	pub(crate) fn new(store: Arc<dyn KvStore>, authorizer: Arc<dyn Authorizer>) -> Self {
+		Self { store, authorizer }
+	}
+}
+
+impl Service<Request<Incoming>> for VssService {
+	type Response = Response<Full<Bytes>>;
+	type Error = hyper::Error;
+	type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+	fn call(&self, req: Request<Incoming>) -> Self::Future {
+		let store = Arc::clone(&self.store);
+		let authorizer = Arc::clone(&self.authorizer);
+		let path = req.uri().path().to_owned();
+		Box::pin(async move {
+			match path.as_str() {
+				"/getObject" => {
+					handle_request(store, authorizer, req, handle_get_object_request).await
+				},
+				"/putObjects" => {
+					handle_request(store, authorizer, req, handle_put_object_request).await
+				},
+				"/deleteObject" => {
+					handle_request(store, authorizer, req, handle_delete_object_request).await
+				},
+				"/listKeyVersions" => {
+					handle_request(store, authorizer, req, handle_list_object_request).await
+				},
+				_ => {
+					let error = format!("Unknown request: {}", path).into_bytes();
+					Ok(Response::builder()
+						.status(StatusCode::BAD_REQUEST)
+						.body(Full::new(Bytes::from(error)))
+						.unwrap())
+				},
+			}
+		})
+	}
+}
+
+async fn handle_get_object_request(
+	store: Arc<dyn KvStore>, user_token: String, request: GetObjectRequest,
+) -> Result<GetObjectResponse, VssError> {
+	store.get(user_token, request).await
+}
+async fn handle_put_object_request(
+	store: Arc<dyn KvStore>, user_token: String, request: PutObjectRequest,
+) -> Result<PutObjectResponse, VssError> {
+	store.put(user_token, request).await
+}
+async fn handle_delete_object_request(
+	store: Arc<dyn KvStore>, user_token: String, request: DeleteObjectRequest,
+) -> Result<DeleteObjectResponse, VssError> {
+	store.delete(user_token, request).await
+}
+async fn handle_list_object_request(
+	store: Arc<dyn KvStore>, user_token: String, request: ListKeyVersionsRequest,
+) -> Result<ListKeyVersionsResponse, VssError> {
+	store.list_key_versions(user_token, request).await
+}
+async fn handle_request<
+	T: Message + Default,
+	R: Message,
+	F: FnOnce(Arc<dyn KvStore>, String, T) -> Fut + Send + 'static,
+	Fut: Future<Output = Result<R, VssError>> + Send,
+>(
+	store: Arc<dyn KvStore>, authorizer: Arc<dyn Authorizer>, request: Request<Incoming>,
+	handler: F,
+) -> Result<<VssService as Service<Request<Incoming>>>::Response, hyper::Error> {
+	// TODO: we should bound the amount of data we read to avoid allocating too much memory.
+	let (parts, body) = request.into_parts();
+	let headers_map = parts
+		.headers
+		.iter()
+		.map(|(k, v)| (k.as_str().to_string(), v.to_str().unwrap_or("").to_string()))
```

Comment:
Should we consider passing the `HeaderMap` instead to avoid these allocations and conversion of non-ASCII values to something unusable?